File: src/main/java/org/tasks/ui/TimePreference.java
package org.tasks.ui;

import android.content.Context;
import android.content.Intent;
import android.content.res.TypedArray;
import android.preference.Preference;
import android.util.AttributeSet;

import com.todoroo.andlib.utility.DateUtilities;

import org.tasks.R;
import org.tasks.activities.TimePickerActivity;
import org.tasks.time.DateTime;

public class TimePreference extends Preference {

    private int millisOfDay;
    private String summary;

    public TimePreference(Context context, AttributeSet attrs) {
        super(context, attrs);

        TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.TimePreference);
        summary = a.getString(R.styleable.TimePreference_summary);
        a.recycle();
    }

    @Override
    protected Object onGetDefaultValue(TypedArray a, int index) {
        return a.getString(index);
    }

    @Override
    public void onSetInitialValue(boolean restoreValue, Object defaultValue) {
        if (restoreValue) {
            int noon = new DateTime().startOfDay().withHourOfDay(12).getMillisOfDay();
            millisOfDay = getPersistedInt(noon);
        } else {
            millisOfDay = Integer.parseInt((String) defaultValue);
        }

        setMillisOfDay(millisOfDay);
    }

    public int getMillisOfDay() {
        return millisOfDay;
    }

    public void handleTimePickerActivityIntent(Intent data) {
        long timestamp = data.getLongExtra(TimePickerActivity.EXTRA_TIMESTAMP, 0L);
        int millisOfDay = new DateTime(timestamp).getMillisOfDay();
        if (callChangeListener(millisOfDay)) {
            persistInt(millisOfDay);
            setMillisOfDay(millisOfDay);
        }
    }

    private void setMillisOfDay(int millisOfDay) {
        this.millisOfDay = millisOfDay;
        String setting = DateUtilities.getTimeString(getContext(), new DateTime().withMillisOfDay(millisOfDay));
        setSummary(summary == null ? setting : String.format(summary, setting));
    }
}

File: app/src/main/java/org/tasks/ui/TimePreference.kt
package org.tasks.ui

import android.content.Context
import android.content.Intent
import android.util.AttributeSet
import androidx.preference.Preference
import org.tasks.R
import org.tasks.dialogs.MyTimePickerDialog
import org.tasks.extensions.Context.is24HourFormat
import org.tasks.kmp.org.tasks.time.getTimeString
import org.tasks.time.DateTime
import org.tasks.time.DateTimeUtils2.currentTimeMillis
import org.tasks.time.withMillisOfDay

class TimePreference(context: Context, attrs: AttributeSet) : Preference(context, attrs) {

    private val summaryTemplate: String?
    var millisOfDay: Int = 0
        private set

    init {
        val a = context.obtainStyledAttributes(attrs, R.styleable.TimePreference)
        summaryTemplate = a.getString(R.styleable.TimePreference_time_summary)
        a.recycle()
    }

    override fun onGetDefaultValue(a: android.content.res.TypedArray, index: Int): Any {
        return a.getInteger(index, -1)
    }

    override fun onSetInitialValue(restoreValue: Boolean, defaultValue: Any?) {
        millisOfDay = if (restoreValue) {
            val noon = DateTime().startOfDay().withHourOfDay(12).millisOfDay
            getPersistedInt(noon)
        } else {
            defaultValue as Int
        }
        updateMillisOfDay(millisOfDay)
    }

    private fun updateMillisOfDay(millis: Int) {
        millisOfDay = millis
        val setting = getTimeString(
            currentTimeMillis().withMillisOfDay(millis),
            context.is24HourFormat
        )
        summary = if (summaryTemplate == null) setting else String.format(summaryTemplate, setting)
    }

    fun handleTimePickerActivityIntent(data: Intent) {
        val timestamp = data.getLongExtra(MyTimePickerDialog.EXTRA_TIMESTAMP, 0L)
        val millis = DateTime(timestamp).millisOfDay
        if (callChangeListener(millis)) {
            persistInt(millis)
            updateMillisOfDay(millis)
        }
    }
}
